Import Map support

wraiford opened the pull request:
**NOTE: This is still a rough draft and is not ready to be merged atow (April 18, 2023)**

## Goal

This PR is to add functionality to enable import maps when testing ES Module libraries in the browser context. The consumer should be able to add config in `jasmine-browser.json` that results in an importmap `<script>` tag in the html template of the browser harness, such that ESM imports resolve correctly and bare imports errors are avoided.

This references issue #28

## Broad overview of changes

* Config entry
* Inject importmap into template
* Add type documentation (jsdoc)
* Associated testing, including a sample consumer project

## links

* [import maps spec on github](https://github.com/WICG/import-maps)
* [importmap on MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script/type/importmap)

sgravrock commented:
Thanks for this. Free time has been scarce this week but I should be able to dig into this and respond to your earlier questions this weekend.

sgravrock reviewed (commented):
I should be able to give this a more thorough review and tire kicking in a day or three, but I wanted to mention some low hanging fruit that I noticed in my initial pass.

In addition to my inline comments: CI builds are failing on Windows and on Node 12. 

The tests are failing on [Node 12](https://app.circleci.com/pipelines/github/jasmine/jasmine-browser-runner/1065/workflows/b24aec42-0623-46e0-9f7f-074554072e84/jobs/5001) because the `??` operator doesn't exist yet, and on [Windows](https://app.circleci.com/pipelines/github/jasmine/jasmine-browser-runner/1065/workflows/b24aec42-0623-46e0-9f7f-074554072e84/jobs/5002) because the import map in the output HTML contains backslashes instead of slashes. You should be able to fix Windows by avoiding `path.sep` and by replacing backslashes with slashes in the result of `path.join`. 

As for Node 12, it's up to you whether you want to support it. I'm planning to release 2.0, which will drop everything below Node 18 and probably require jasmine-core 5.x, in a month or two. So you've got two choices:

* Convert to Node 12 compatible syntax. If this PR merges before 2.0, I'll release it in 1.4.0.
* Change the PR's destination branch to 2.0. It'll be released in 2.x.

I'm fine with either approach.

wraiford commented on spec/serverSpec.js:
```diff
@@ -520,4 +520,106 @@ describe('server', function() {
+      await this.server.stop();
+    });
+
+    it('should start server with inline import map in html', async function() {
```
I hear you on this one. I'm not sure if there's really a right answer, but this is the kind of sovereign opinion that you just have to tell me what you want. At the time, I was just trying to get something working and I certainly appreciate clarity (as I said, I use specs for learning APIs). I will see how much it blows up if I try to reduce DRY.

wraiford replied:
I've tried an iteration that breaks out these specs to be more concrete, less DRY, and ultimately hopefully more readable in [this commit](https://github.com/jasmine/jasmine-browser-runner/pull/30/commits/359331a72663e069d71ef9f8f1623add2a2ac4a5).

wraiford commented on spec/configSpec.js:
```diff
@@ -118,5 +118,185 @@ describe('config', function() {
+          config['importMap'] = {};
+          validateConfig(config);
+        }).toThrowError(
+          /Configuration.+importMap.+contains no imports or scopes/
```
Again, your opinion rules here. Personally, I've become less of a fan of matching against string literals over the years and I thought it was close enough visually and effectively to the existing ones. But I'll go ahead and change it over if you feel comfortable with all of the error messages (those messages could have been something you wanted to change in code review for example, but I couldn't know your opinion on those beforehand!).

wraiford replied:
I've changed this to string literals in [this commit](https://github.com/jasmine/jasmine-browser-runner/pull/30/commits/27b1e96c31ead16a991dc93a7949b4fa2d25f79e).

Additionally, this only took a couple minutes...but it was a painful couple of minutes. This was precisely the kind of stylistic opinion I would have appreciated you giving feedback on earlier to avoid having to do this. But on the other hand, I've skimmed over some of the neat stuff you've been doing in the pivotal tracker backlog and I can definitely see how you might not have had the time to notice this and form this opinion yourself before.

And to be a bit cheeky (though inflection is lost in text-only comms and we don't really know each other), no I don't think it would be simpler or clearer, though possibly more effective on a large scale but not at this scale. Personally I use UUIDs (or GUIDs depending on your religion!) because I think the overhead is worth the ease with which you can just search in a codebase for a unique sequence of characters. Otherwise, you have to parse the error message and hope someone hasn't copy/pasted too large of a portion of the error message, let alone those errors that are exactly the same strings! That said, of course I don't see you retrofitting a hugely popular and already well-established codebase to use these and as you can see from the commit, I defer to your sovereign rule here! 👑

sgravrock replied:
> Additionally, this only took a couple minutes...but it was a painful couple of minutes. This was precisely the kind of stylistic opinion I would have appreciated you giving feedback on earlier to avoid having to do this. 

I provided that feedback as soon as I saw the code in question, which I *think* is also the first time I had access to it. It's possible that there was a spec like this in one of the commits you linked to prior to opening the PR and I missed it. But in any case, I don't generally give WIP that kind of thorough, detail-oriented review. For two reasons: because I usually don't have enough time to give PRs multiple passes at that level of detail, and because a lot of developers don't appreciate getting that type of feedback when they're still trying to get things working.

What I *did* do is write all of the existing config validation specs in a consistent style. In a way, that *is* an expression of stylistic opinion. You could have avoided the rework by matching the existing code rather than solving the same problem in a different way.

I appreciate that you've ultimately been willing to conform to the local style. But there have been several cases where you've done that only after departing from it and advocating for your way of doing things. That can be ok (although maintainer reactions to it can vary greatly) but it's not a good strategy if you're looking to get a PR merged with a minimum of change requests.

That's not to say that you can't write the code the way you want to. You don't owe anyone a PR at all, let alone one that's written in a particular style. But keep in mind that "I want to do things my way" and "I want to get this merged with a minimum of fuss and rework" tend to be mutually exclusive goals that call for different mindsets.

wraiford commented:
> The tests are failing on [Node 12](https://app.circleci.com/pipelines/github/jasmine/jasmine-browser-runner/1065/workflows/b24aec42-0623-46e0-9f7f-074554072e84/jobs/5001) because the `??` operator doesn't exist yet...

I've changed to a logical OR [with this commit](https://github.com/jasmine/jasmine-browser-runner/pull/30/commits/d1d4d65bfc7d7c7c84e32fd3856daef5c7dc4355).

wraiford commented on lib/server.js:
```diff
@@ -115,6 +116,30 @@ class Server {
     return [].concat(srcUrls, helperUrls, specUrls);
   }
 
+  /**
```
I've changed this to slashed comments in [this commit](https://github.com/jasmine/jasmine-browser-runner/pull/30/commits/55c5ba9f9448c2a1c1d983299b1c218381959607).

wraiford commented:
Ah well, I tried a blind stab at the windows fix. I'll have to get it setup on my windows machine, which I've only just gotten going recently. I don't even have vscode on it yet, and I don't have any windows VMs going atm either (I do very little in VMs these days...getting old!).

sgravrock commented:
> Ah well, I tried a blind stab at the windows fix. I'll have to get it setup on my windows machine, which I've only just gotten going recently. I don't even have vscode on it yet, and I don't have any windows VMs going atm either (I do very little in VMs these days...getting old!

Don’t worry about it. I thought the blind stab would work, but since it didn’t, I can take care of Windows. I already have a test machine handy.

wraiford commented:
> Don’t worry about it. I thought the blind stab would work, but since it didn’t, I can take care of Windows. I already have a test machine handy.

Well I didn't worry about it but I did go ahead and get a VM up and going! (I already had a partially initialized Windows VM after all) For me personally, it was throwing an error about not finding ChromeDriver:

> Failed: Child suite failed with error:
>     Error: Command failed: node ../../../bin/jasmine-browser-runner runSpecs
>     Error: The ChromeDriver could not be found on the current PATH.
>           Please download the latest version of the ChromeDriver
>           from http://chromedriver.storage.googleapis.com/index.html
>           and ensure it can be found on your PATH.
>         at new ServiceBuilder (C:\Users\billm\ibgib\forks\jasmine-browser-runner\node_modules\selenium-webdriver\chrome.js:175:13)
>         at thenableWebDriverProxy.getDefaultService (C:\Users\billm\ibgib\forks\jasmine-browser-runner\node_modules\selenium-webdriver\chrome.js:264:12)
>         at thenableWebDriverProxy.createSession (C:\Users\billm\ibgib\forks\jasmine-browser-runner\node_modules\selenium-webdriver\chromium.js:669:49)
>         at thenableWebDriverProxy.createSession (C:\Users\billm\ibgib\forks\jasmine-browser-runner\node_modules\selenium-webdriver\chrome.js:255:13)
>         at createDriver (C:\Users\billm\ibgib\forks\jasmine-browser-runner\node_modules\selenium-webdriver\index.js:147:33)
>         at Builder.build (C:\Users\billm\ibgib\forks\jasmine-browser-runner\node_modules\selenium-webdriver\index.js:675:16)
>         at buildWebdriver (C:\Users\billm\ibgib\forks\jasmine-browser-runner\lib\webdriver.js:31:10)
>         at Object.runSpecs (C:\Users\billm\ibgib\forks\jasmine-browser-runner\index.js:101:25)
>         at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>         at async Command.runSpecs (C:\Users\billm\ibgib\forks\jasmine-browser-runner\lib\command.js:187:5)
> 
> 
>     stdout:
>     Jasmine server is running here: http://localhost:50624
>     Jasmine tests are here:         C:\Users\billm\ibgib\forks\jasmine-browser-runner\spec\fixtures\esmIntegration\spec
>     Source files are here:          C:\Users\billm\ibgib\forks\jasmine-browser-runner\spec\fixtures\esmIntegration\src
>      The ChromeDriver could not be found on the current PATH, trying Selenium Manager
>     Unable to obtain driver using Selenium Manager: Error: Error executing command with C:\Users\billm\ibgib\forks\jasmine-browser-runner\node_modules\selenium-webdriver\bin\windows\selenium-manager.exe,--browser,chrome
> 
> 
> 
>     stderr:
>     Error: The ChromeDriver could not be found on the current PATH.
>           Please download the latest version of the ChromeDriver
>           from http://chromedriver.storage.googleapis.com/index.html
>           and ensure it can be found on your PATH.
>         at new ServiceBuilder (C:\Users\billm\ibgib\forks\jasmine-browser-runner\node_modules\selenium-webdriver\chrome.js:175:13)
>         at thenableWebDriverProxy.getDefaultService (C:\Users\billm\ibgib\forks\jasmine-browser-runner\node_modules\selenium-webdriver\chrome.js:264:12)
>         at thenableWebDriverProxy.createSession (C:\Users\billm\ibgib\forks\jasmine-browser-runner\node_modules\selenium-webdriver\chromium.js:669:49)
>         at thenableWebDriverProxy.createSession (C:\Users\billm\ibgib\forks\jasmine-browser-runner\node_modules\selenium-webdriver\chrome.js:255:13)
>         at createDriver (C:\Users\billm\ibgib\forks\jasmine-browser-runner\node_modules\selenium-webdriver\index.js:147:33)
>         at Builder.build (C:\Users\billm\ibgib\forks\jasmine-browser-runner\node_modules\selenium-webdriver\index.js:675:16)
>         at buildWebdriver (C:\Users\billm\ibgib\forks\jasmine-browser-runner\lib\webdriver.js:31:10)
>         at Object.runSpecs (C:\Users\billm\ibgib\forks\jasmine-browser-runner\index.js:101:25)
>         at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
>         at async Command.runSpecs (C:\Users\billm\ibgib\forks\jasmine-browser-runner\lib\command.js:187:5)

I tried installing `chromedriver` globally and still no dice, even though my nodejs folder (which was indeed on the PATH) contained files `chromedriver` and `chromedriver.cmd`. I downloaded the latest chromedriver win32 zip [here](http://chromedriver.storage.googleapis.com/index.html?path=113.0.5672.63/) and extracted the `chromedriver.exe` to the nodejs folder, updated my chrome browser proper (chromedriver version has to match), and the specs passed except the pending sauce labs spec.

Though there were quite a few logs that I don't see in mac in the form of:

>  [0503/130519.036:INFO:CONSOLE(3)] "spec 4", source: http://localhost:51412/__spec__/aSpec.js (3)
> [0503/130519.039:INFO:CONSOLE(159)] "Spec '4' has no expectations.", source: http://localhost:51412/__jasmine__/jasmine-html.js (159)

## notes

* This was testing both the existing esmintegration spec (via `fdescribe`), and the specs I wrote.
* I installed node via nvm-windows
  * I tested both v14 & v20 with identical chromedriver errors.
  * As an aside about nvm-windows, I had to copy `chromedriver.exe` to nodejs after switching versions using `nvm use` but the `chromedriver.exe` file persisted when I switched between the two versions subsequently. 
* Looking at the circleci details (which I just noticed 🙄) it is failing on the regexp matching. 
  * This is not failing on my local testing.
  * I visually checked to ensure I have the most recent version that has the more extensive expression checking.
  * I also tested with the exact node/npm versions in the circleci build of "Node v14.17.5 and NPM 6.14.14" and the specs are passing.

## summary

So for me it seems the specs themselves are passing, but not on the build server. It looks like it's a Windows Server image, whereas I'm on Windows Home. I'm not sure how that might make either the html output be materially different or if the node version's regular expression engine somehow differ. 🤷 

I can dumb down the regular expression checking and just check for individual lines, or if you have any suggestions on writing a better regular expression, I'm all ears (I'm not a pro at them but usually I can get them working).

wraiford commented:
I thought I saw a difference in the length of texts. Now with them as `RegExp` objects, the output error shows them as regular expressions whose source's are the same length. However, it still failed with regular expressions. 

However!

In the very same file, there are multiple comments about problems with line endings of `\r?\n`, which was in the back of my mind...that and the fact that line endings are a pita often. Now it looks to be passing, having removed the `\n` which was redundant anyway since that matches against the `\s*` part. 😅

wraiford commented:
So I await your kicking the hell out of those tires.

sgravrock commented:
This is merged in aabd966, but GitHub doesn't show it as merged because I squashed the commits. Thanks for all your hard work on this.

wraiford commented:
Hopefully it will benefit many users of a great library! Thanks for working with me 👍 😃
